packages/pubsub/src/subscription.js

```diff
@@ -850,114 +688,7 @@ Subscription.prototype.setAckDeadline = function(options, callback) {
- * how many message listeners are assigned, and then removed, making sure
- * polling is handled automatically.
- *
- * As long as there is one active message listener, the connection is open. As
- * soon as there are no more message listeners, the connection is closed.
- *
- * @private
- *
- * @example
- * subscription.listenForEvents_();
- */
-Subscription.prototype.listenForEvents_ = function() {
-  var self = this;
-
-  this.on('newListener', function(event) {
-    if (event === 'message') {
-      self.messageListeners++;
-      if (self.closed) {
-        self.closed = false;
-        self.startPulling_();
-      }
-    }
-  });
-
-  this.on('removeListener', function(event) {
-    if (event === 'message' && --self.messageListeners === 0) {
-      self.closed = true;
-
-      if (self.activeRequest_ && self.activeRequest_.abort) {
-        self.activeRequest_.abort();
-      }
-    }
-  });
-};
-
-/**
- * Update the status of `maxInProgress`. Å subscription becomes "paused" (not
- * pulling) when the number of messages that have yet to be ack'd or skipped
- * exceeds the user's specified `maxInProgress` value.
- *
- * This will start pulling when that event reverses: we were paused, but one or
- * more messages were just ack'd or skipped, freeing up room for more messages
- * to be consumed.
- *
- * @private
- */
-Subscription.prototype.refreshPausedStatus_ = function() {
-  var isCurrentlyPaused = this.paused;
-  var inProgress = Object.keys(this.inProgressAckIds).length;
-
-  this.paused = inProgress >= this.maxInProgress;
-
-  if (isCurrentlyPaused && !this.paused && this.messageListeners > 0) {
-    this.startPulling_();
-  }
-};
-
-/**
- * Poll the backend for new messages. This runs a loop to ping the API at the
- * provided interval from the subscription's instantiation. If one wasn't
- * provided, the default value is 10 milliseconds.
- *
- * If messages are received, they are emitted on the `message` event.
- *
- * Note: This method is automatically called once a message event handler is
- * assigned to the description.
- *
- * To stop pulling, see {module:pubsub/subscription#close}.
- *
- * @private
- *
- * @example
- * subscription.startPulling_();
- */
-Subscription.prototype.startPulling_ = function() {
-  var self = this;
-
-  if (this.closed || this.paused) {
-    return;
-  }
-
-  var maxResults;
-
-  if (this.maxInProgress < Infinity) {
-    maxResults = this.maxInProgress - Object.keys(this.inProgressAckIds).length;
-  }
-
-  this.pull({
-    returnImmediately: false,
-    maxResults: maxResults
-  }, function(err, messages, apiResponse) {
-    if (err) {
-      self.emit('error', err, apiResponse);
-    }
-
-    if (messages) {
-      messages.forEach(function(message) {
-        self.emit('message', message, apiResponse);
-      });
-    }
-
-    setTimeout(self.startPulling_.bind(self), self.interval);
-  });
+  return this.pubsub.snapshot.call(this, name);
 };
 
```

Comment:
@callmehiphop @jganetsk 

First, we want to use a 99th percentile algorithm for determining how frequently to modify the ack deadlines and what to modify them to. In other words, every time we ack a message, we keep track of the time delta between when we got the message and when we acked it, and add that to a histogram. We can check the histogram at any point to see what the 99th percentile is (e.g. the duration such that only 1% of the messages took longer than this to ack).

I have a work-in-progress implementation of this concept in Python [here][1]. Ideally I would like the Ruby and Node clients to work the same way.

Beyond implementing the histogram itself, you basically need to check the histogram to know how often to modify the ack deadlines. So the implementation will end up looking like this:

  * When the subscriber first starts, set an `ackDeadline` value, initialized to 10,000 (10 seconds).
  * After `Math.random() * ackDeadline * 0.9` seconds, do the following:
      * Check the histogram for the new p99 value.
      * Modify the ack deadline of all outstanding messages to the p99 value.
      * Set the `ackDeadline` variable to the new p99 value.
  * Repeat previous step over and over.

  [1]: https://github.com/GoogleCloudPlatform/google-cloud-python/blob/pubsub-subscriber/pubsub/google/cloud/pubsub_v1/subscriber/histogram.py